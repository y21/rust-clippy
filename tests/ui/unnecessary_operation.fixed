#![allow(
    clippy::deref_addrof,
    dead_code,
    unused,
    clippy::no_effect,
    clippy::unnecessary_struct_initialization
)]
#![warn(clippy::unnecessary_operation)]

struct Tuple(i32);
struct Struct {
    field: i32,
}
enum Enum {
    Tuple(i32),
    Struct { field: i32 },
}
struct DropStruct {
    field: i32,
}
impl Drop for DropStruct {
    fn drop(&mut self) {}
}
struct DropTuple(i32);
impl Drop for DropTuple {
    fn drop(&mut self) {}
}
enum DropEnum {
    Tuple(i32),
    Struct { field: i32 },
}
impl Drop for DropEnum {
    fn drop(&mut self) {}
}
struct FooString {
    s: String,
}

fn get_number() -> i32 {
    0
}

fn get_usize() -> usize {
    0
}
fn get_struct() -> Struct {
    Struct { field: 0 }
}
fn get_drop_struct() -> DropStruct {
    DropStruct { field: 0 }
}

fn main() {
    get_number();
    get_number();
    get_struct();
    get_number();
    get_number();
    5;get_number();
    get_number();
    get_number();
    5;6;get_number();
    get_number();
    get_number();
    5;get_number();
    42;get_number();
    assert!([42, 55].len() > get_usize());
    42;get_number();
    get_number();
    assert!([42; 55].len() > get_usize());
    get_number();
    String::from("blah");

    // Do not warn
    DropTuple(get_number());
    DropStruct { field: get_number() };
    DropStruct { field: get_number() };
    DropStruct { ..get_drop_struct() };
    DropEnum::Tuple(get_number());
    DropEnum::Struct { field: get_number() };

    // Issue #9954
    fn one() -> i8 {
        1
    }
    macro_rules! use_expr {
        ($($e:expr),*) => {{ $($e;)* }}
    }
    use_expr!(isize::MIN / -(one() as isize), i8::MIN / -one());
}

fn issue9951() {
    {
        // Original repro
        #[derive(Clone, Default)]
        struct MaybeCopy<T>(T);

        impl Copy for MaybeCopy<u8> {}

        // Since `MaybeCopy` is generic and there is only one `Copy` impl for it that doesn't use type
        // parameters, and array repeat expression requires the type implement `Copy`,
        // type inference can deduce that `MaybeCopy<?>` must be `MaybeCopy<u8>`.
        // However if we were to reduce it to just `MaybeCopy::default()`, there wouldn't be a `Copy`
        // constraint anymore, creating compile errors. Don't lint here!
        [MaybeCopy::default(); 13];

        // `Copy` is not required in array repeat expressions when array length is 0 or 1.
        // In that case, reducing it is ok because there's no such `Copy` constraint that would help
        // type inference here.
        MaybeCopy::<u8>::default();
        //~^ ERROR: unnecessary operation
        MaybeCopy::<u8>::default();
        //~^ ERROR: unnecessary operation
    }

    {
        #[derive(Default, Clone)]
        struct S<T>(T);

        impl<T: Copy> Copy for S<T> {}

        S::<u8>::default();
        //~^ ERROR: unnecessary operation
        S::<u8>::default();
        //~^ ERROR: unnecessary operation
    }

    {
        #[derive(Default, Clone)]
        struct S<T>(T);

        impl Copy for S<u8> {}
        impl Copy for S<u16> {}

        // Similar to a generic `S<T>: Copy` impl, there are multiple candidates, so we can lint.
        S::<u8>::default();
        //~^ ERROR: unnecessary operation
        S::<u8>::default();
        //~^ ERROR: unnecessary operation
    }

    {
        #[derive(Default, Clone)]
        struct S<T>(T);
        #[derive(Default, Clone)]
        struct S2<T>(T);

        impl<T: Copy> Copy for S<T> {}
        // Nested type parameter
        impl<T: Copy> Copy for S2<S<T>> {}

        S2::<S<u8>>::default();
        //~^ ERROR: unnecessary operation
        S2::<S<u8>>::default();
        //~^ ERROR: unnecessary operation
    }
}
