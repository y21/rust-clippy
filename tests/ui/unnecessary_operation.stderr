error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:54:5
   |
LL |     Tuple(get_number());
   |     ^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`
   |
   = note: `-D clippy::unnecessary-operation` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unnecessary_operation)]`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:55:5
   |
LL |     Struct { field: get_number() };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:56:5
   |
LL |     Struct { ..get_struct() };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_struct();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:57:5
   |
LL |     Enum::Tuple(get_number());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:58:5
   |
LL |     Enum::Struct { field: get_number() };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:59:5
   |
LL |     5 + get_number();
   |     ^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `5;get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:60:5
   |
LL |     *&get_number();
   |     ^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:61:5
   |
LL |     &get_number();
   |     ^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:62:5
   |
LL |     (5, 6, get_number());
   |     ^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `5;6;get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:63:5
   |
LL |     get_number()..;
   |     ^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:64:5
   |
LL |     ..get_number();
   |     ^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:65:5
   |
LL |     5..get_number();
   |     ^^^^^^^^^^^^^^^^ help: statement can be reduced to: `5;get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:66:5
   |
LL |     [42, get_number()];
   |     ^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `42;get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:67:5
   |
LL |     [42, 55][get_usize()];
   |     ^^^^^^^^^^^^^^^^^^^^^^ help: statement can be written as: `assert!([42, 55].len() > get_usize());`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:68:5
   |
LL |     (42, get_number()).1;
   |     ^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `42;get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:69:5
   |
LL |     [get_number(); 55];
   |     ^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:70:5
   |
LL |     [42; 55][get_usize()];
   |     ^^^^^^^^^^^^^^^^^^^^^^ help: statement can be written as: `assert!([42; 55].len() > get_usize());`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:71:5
   |
LL | /     {
LL | |         get_number()
LL | |     };
   | |______^ help: statement can be reduced to: `get_number();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:74:5
   |
LL | /     FooString {
LL | |         s: String::from("blah"),
LL | |     };
   | |______^ help: statement can be reduced to: `String::from("blah");`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:114:9
   |
LL |         [MaybeCopy::<u8>::default(); 0];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `MaybeCopy::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:116:9
   |
LL |         [MaybeCopy::<u8>::default(); 1];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `MaybeCopy::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:126:9
   |
LL |         [S::<u8>::default(); 13];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:128:9
   |
LL |         [S::<u8>::default(); 0];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:140:9
   |
LL |         [S::<u8>::default(); 13];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:142:9
   |
LL |         [S::<u8>::default(); 0];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S::<u8>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:156:9
   |
LL |         [S2::<S<u8>>::default(); 13];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S2::<S<u8>>::default();`

error: unnecessary operation
  --> $DIR/unnecessary_operation.rs:158:9
   |
LL |         [S2::<S<u8>>::default(); 0];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `S2::<S<u8>>::default();`

error: aborting due to 27 previous errors

